### PR TITLE
feat: add packet provisioner

### PR DESCRIPTION
**Pull request by rsmitty**
Will close out #10

### CODE REVIEW

File: config/samples/cluster-deployment/packet/platform-config-masters.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        plan: "t1.small.x86"
+        facility: "sjc1"
+        pxeurl: "http://{{PXE_SERVER}}/boot.ipxe"
+        install:
+          wipe: false
+          force: true
+          boot:
+            device: /dev/sda
+            size: 1024000000
+            kernel: http://{{PXE_SERVER}}/assets/talos/latest/vmlinuz
```
Comment (bradbeam) [inline config/samples/cluster-deployment/packet/platform-config-masters.yaml]: think it's worth adding a comment here around these urls being optional?

File: pkg/cloud/talos/provisioners/packet/packet.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+	}
+	udStruct := &Userdata{}
+	yaml.Unmarshal([]byte(udConfigMap.Data["userdata"]), udStruct)
+
+	if packetConfig.Instances.Install != nil {
+		udStruct.Install = packetConfig.Instances.Install
+	}
+
+	//Add network tweaks for elastic IPs to userdata
+	floatingIP := ""
```
Comment (bradbeam) [inline pkg/cloud/talos/provisioners/packet/packet.go]: I think they're both equivalent, but typically we'd do 
```
var floatingIP string
```
in this case.

File: pkg/cloud/talos/provisioners/packet/packet.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+	isMaster := strings.Contains(machine.ObjectMeta.Name, "master")
+	if isMaster {
+		nameSlice := strings.Split(machine.ObjectMeta.Name, "-")
+		indexString := nameSlice[len(nameSlice)-1]
+		index, err := strconv.Atoi(indexString)
+		if err != nil {
+			return err
+		}
+		floatingIP = clusterSpec.Masters.IPs[index]
```
Comment (bradbeam) [inline pkg/cloud/talos/provisioners/packet/packet.go]: This section feels pretty fragile and I think will fail if we try to do a rollout strategy of +1 node, -1 node while trying to preserve the master IPs. I'm not sure if that's a use case we are trying to cover, but wanted to mention. 

As far as suggestions/alternatives/workarounds I'm not sure I've got enough context of the above with what's available to try to decide off of. Maybe we can chat more about it.

File: pkg/cloud/talos/provisioners/packet/packet.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	}
+
+	//Wait for masters to be active, attach floating ip
+	if isMaster {
+		err = packet.waitForStatus(machine, "active")
+		if err != nil {
+			return err
+		}
+		ipReq := &packngo.AddressStruct{Address: floatingIP + "/32"}
+		_, _, err = packet.client.DeviceIPs.Assign(dev.ID, ipReq)
```
Comment (bradbeam) [inline pkg/cloud/talos/provisioners/packet/packet.go]: Do we need to do a detach event or will the above handle that automatically?

File: pkg/cloud/talos/provisioners/packet/packet.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+	return nil, nil
+}
+
+//waitForStatus polls the Packet api for a certain instance status
+//needed for attaching elastic IP after boot
+func (packet *Packet) waitForStatus(machine *clusterv1.Machine, desiredState string) error {
+
+	timeout := time.After(300 * time.Second)
```
Comment (bradbeam) [inline pkg/cloud/talos/provisioners/packet/packet.go]: I think it's worth increasing this timeout a bit more, probably at least 600s to be safe. Usually things are pretty quick, but I've run into instances where you need to wait a few moments/minutes for devices to be returned to the pool before provisioning can begin.

File: pkg/cloud/talos/provisioners/packet/packet.go (hunk 3 above)
Comment (rsmitty) [reply pkg/cloud/talos/provisioners/packet/packet.go]: Added extra check for index vs. IP list

File: pkg/cloud/talos/provisioners/packet/packet.go (hunk 5 above)
Comment (rsmitty) [reply pkg/cloud/talos/provisioners/packet/packet.go]: Fixed

File: pkg/cloud/talos/provisioners/packet/packet.go (hunk 2 above)
Comment (rsmitty) [reply pkg/cloud/talos/provisioners/packet/packet.go]: Fixed

File: pkg/cloud/talos/provisioners/packet/packet.go (hunk 4 above)
Comment (rsmitty) [reply pkg/cloud/talos/provisioners/packet/packet.go]: Need to explore this. Will create backlog ticket to make sure the floating IP just gets move automatically. The way the docs are written, I feel like it should.

**Pull request merged**